### PR TITLE
feature(rslint_parser): Fix grammar mismatches/add empty slots

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

I started working on a prototype that implements the schema verification of our AST nodes as described in #1867 and noticed that we still have many mismatches between the grammar or missing empty slots. 

This PR fixes many of these mismatches. There are certainly more and refactoring `expr()` will help us to discover many of them. This PR addresses all (excluding classes) that are currently part of our integration tests. 

* Remove recovery from `parse_statement` and return `ParsedSyntax`
* Fix missing markers of parenthesized expressions
* Remove `missing` marker for computed member assignments
* Add missing marker for `missing` return type annotations of object members
* Add missing marker for `async` and type parameters of arrow functions
* Add missing markers for the argument and `*` token of yield expressions
* Add missing markers for missing `left` and right of binary expressions
* Add missing marker for the call arg type arguments
* Add missing markers for the optional chain operator and expression of computed member expressions
* Fix ordering of typ eargs for new and call expr
* Add missing marker for missing throw and return argument
* Add missing marker for missing parenthesized expressions
* Add missing marker for missing if/else/while/do body
* Fix position of `type_args` for async arrow function expressions
* Remove `hole_token` from `JsArrayHole`
* Add the EOF token to JsScript & JsModule. We always parsed it but didn't add it to the AST Facade.
* Binding / Assignment target with default: Add a missing marker for a missing initializer.
* Directives: Add a missing marker for missing semicolons
* Functions: Add missing markers for missing "async" and "*" tokens

Parat of #1725 and #1724 

## Test Plan

Verified Rast changes, made sure that nodes are "valid" on the branch that verifies the AST node correctness.

### CODE REVIEW

File: crates/rslint_parser/src/parser.rs [hunk 1]
```diff
@@ -726,6 +730,35 @@ impl CompletedMarker {
...
+	/// The position of this marker in the `positions` array. Guaranteed
+	/// to point to an `Event::Missing`
+	events_pos: u32,
+	/// The tokens position at the moment the missing slot was created
+	token_pos: usize,
+}
+
+impl CompletedMissingMarker {
+	fn new(events_pos: u32, token_pos: usize) -> Self {
+		CompletedMissingMarker {
```
Comment (ematipico) [inline crates/rslint_parser/src/parser.rs:747]:
```suggestion
		Self {
```

File: crates/rslint_parser/src/parser/parsed_syntax.rs [hunk 2]
```diff
@@ -1,12 +1,53 @@
...
+#[derive(Debug, Eq, PartialEq)]
+pub enum CompletedNodeOrMissingMarker {
+	/// Marker for a completed node
+	NodeMarker(CompletedMarker),
+	/// Marker for a completed missing/empty slot
+	MissingMarker(CompletedMissingMarker),
+}
+
+impl CompletedNodeOrMissingMarker {
+	/// Returns `true` if this is a [MissingMarker]
```
Comment (ematipico) [inline crates/rslint_parser/src/parser/parsed_syntax.rs:20]:
```suggestion
	/// Returns `true` if this is a [CompletedNodeOrMissingMarker::MissingMarker]
```

File: crates/rslint_parser/src/parser/parsed_syntax.rs [hunk 3]
```diff
@@ -1,12 +1,53 @@
...
+	pub fn is_missing_marker(&self) -> bool {
+		matches!(self, MissingMarker(_))
+	}
+
+	/// Returns `true` if this is a [NodeMarker]
```
Comment (ematipico) [inline crates/rslint_parser/src/parser/parsed_syntax.rs:25]:
```suggestion
	/// Returns `true` if this is a [CompletedNodeOrMissingMarker::NodeMarker]
```

File: crates/rslint_parser/src/parser/parsed_syntax.rs [hunk 4]
```diff
@@ -1,12 +1,53 @@
...
+	pub fn is_node_marker(&self) -> bool {
+		matches!(self, NodeMarker(_))
+	}
+
+	/// Returns the kind of the completed node if this is a [NodeMarker] or [None]
```
Comment (ematipico) [inline crates/rslint_parser/src/parser/parsed_syntax.rs:30]:
```suggestion
	/// Returns the kind of the completed node if this is a [CompletedNodeOrMissingMarker::NodeMarker] or [None]
```

File: crates/rslint_parser/src/syntax/assignment.rs [hunk 5]
```diff
@@ -378,6 +379,14 @@ impl RewriteParseEvents for ReparseAssignment {
...
 		p.bump_remap(kind);
 	}
+
+	fn missing(&mut self, p: &mut Parser) {
+		if self.parents.last().map(|(parent, _)| parent) == Some(&JS_COMPUTED_MEMBER_ASSIGNMENT) {
+			// computed member assignments can't have optional chaining, drop the missing
+		} else {
+			p.missing();
+		}
+	}
```
Comment (ematipico) [inline crates/rslint_parser/src/syntax/assignment.rs:389]: Having an empty `if` block doesn't make much sense. Wouldn't be better to have a `!=` instead?

**Review by ematipico — APPROVED** [review]
We were really missing a lot of missing markers 🤣

File: crates/rslint_parser/src/syntax/assignment.rs (hunk 5 above)
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/assignment.rs:389]: I like it to have the comment in there. But yeah, I can change it.

**Pull request merged**